File: web/src/engine/gameEngine.ts
import { freshCards } from "./deck";
import { newBonusTokenBank, drawBonus } from "./bonusTokenBank";
import { newTokenBank, exhaustedStackCount, takeTokens } from "./tokenBank";
import { defaultRandom, shuffled, type RandomSource } from "./rng";
import {
  GameError,
  HAND_LIMIT,
  findPlayer,
  isGameOver,
  isRoundOver,
  newPlayer,
  roundBonusValue,
  roundGoodsValue,
  roundResultTotal,
  type Card,
  type GameAction,
  type GameState,
  type Player,
  type RoundResult
} from "./types";

/**
 * Pure, deterministic Jaipur rules engine. `GameState` is treated as an
 * immutable value: every function here takes a state and returns a brand
 * new one (via `structuredClone`) rather than mutating its input. That's
 * what makes it safe to run inside a Firestore transaction (read state,
 * apply, write state) and easy to unit test.
 */

/**
 * Deals the next 5 cards off `drawPile` as a starting hand. Any camels among
 * them go straight to the herd, never into the hand - camels are never
 * treated as ordinary hand cards at any point in the game, and the initial
 * deal is the one place that distinction isn't already enforced by an
 * action's own rules (every in-game action that could add a camel to a
 * hand is rejected by `apply`).
 */
function dealStartingHand(drawPile: Card[]): { hand: Card[]; camelCount: number; remaining: Card[] } {
  const dealt = drawPile.slice(0, 5);
  const remaining = drawPile.slice(5);
  const hand = dealt.filter((card) => card.good !== "camel");
  const camelCount = dealt.length - hand.length;
  return { hand, camelCount, remaining };
}

function dealMarketAndDrawPile(random: RandomSource): { market: Card[]; drawPile: Card[] } {
  let deck = shuffled(freshCards(), random);

  const market: Card[] = [];
  let camelsPlaced = 0;
  deck = deck.filter((card) => {
    if (card.good === "camel" && camelsPlaced < 3) {
      market.push(card);
      camelsPlaced += 1;
      return false;
    }
    return true;
  });
  market.push(...deck.slice(0, 2));
  deck = deck.slice(2);
  return { market, drawPile: deck };
}

export function newGame(
  playerID1: string,
  playerName1: string,
  playerID2: string,
  playerName2: string,
  random: RandomSource = defaultRandom
): GameState {
  const players: [Player, Player] = [newPlayer(playerID1, playerName1), newPlayer(playerID2, playerName2)];
  const { market, drawPile } = dealMarketAndDrawPile(random);

  const deal0 = dealStartingHand(drawPile);
  players[0].hand = deal0.hand;
  players[0].camelCount = deal0.camelCount;
  const deal1 = dealStartingHand(deal0.remaining);
  players[1].hand = deal1.hand;
  players[1].camelCount = deal1.camelCount;
  const remaining = deal1.remaining;

  const startingPlayer = random() < 0.5 ? playerID1 : playerID2;

  return {
    players,
    market,
    drawPile: remaining,
    discard: [],
    tokenBank: newTokenBank(),
    bonusTokenBank: newBonusTokenBank(random),
    currentPlayerID: startingPlayer,
    roundNumber: 1,
    roundEndReason: null,
    lastRoundResults: [],
    winnerID: null
  };
}

/** House rule: the player who scored lower in the previous round opens the
 * next one; on an exact tie, starting player alternates by round number. */
function nextRoundStarter(state: GameState): string {
  if (state.lastRoundResults.length !== 2) {
    return state.players[state.roundNumber % 2].id;
  }
  const [a, b] = state.lastRoundResults;
  const totalA = roundResultTotal(a);
  const totalB = roundResultTotal(b);
  if (totalA === totalB) {
    return state.players[state.roundNumber % 2].id;
  }
  return totalA < totalB ? a.playerID : b.playerID;
}

/** Resets the board for a new round while keeping each player's `roundsWon` tally. */
export function startNextRound(input: GameState, random: RandomSource = defaultRandom): GameState {
  const state = structuredClone(input);
  const starter = nextRoundStarter(state);

  const players = state.players.map(
    (player): Player => ({
      ...player,
      hand: [],
      camelCount: 0,
      wonTokens: {},
      wonBonusTokens: [],
      revealedBonusTokenIndices: []
    })
  ) as [Player, Player];

  const { market, drawPile } = dealMarketAndDrawPile(random);
  const deal0 = dealStartingHand(drawPile);
  players[0].hand = deal0.hand;
  players[0].camelCount = deal0.camelCount;
  const deal1 = dealStartingHand(deal0.remaining);
  players[1].hand = deal1.hand;
  players[1].camelCount = deal1.camelCount;
  const remaining = deal1.remaining;

  return {
    players,
    market,
    drawPile: remaining,
    discard: [],
    tokenBank: newTokenBank(),
    bonusTokenBank: newBonusTokenBank(random),
    currentPlayerID: starter,
    roundNumber: state.roundNumber + 1,
    roundEndReason: null,
    lastRoundResults: [],
    winnerID: null
  };
}

/**
 * Flips one of `playerID`'s bonus tokens face-up. Deliberately not gated by
 * turn order or round/game-over state - either player can flip either
 * player's tokens at any point once they're won, matching the physical
 * game's simultaneous reveal at scoring time. Idempotent: flipping an
 * already-revealed token is a no-op.
 */
export function revealBonusToken(input: GameState, playerID: string, tokenIndex: number): GameState {
  const state = structuredClone(input);
  const player = state.players.find((p) => p.id === playerID);
  if (!player || tokenIndex < 0 || tokenIndex >= player.wonBonusTokens.length) return state;
  if (!player.revealedBonusTokenIndices.includes(tokenIndex)) {
    player.revealedBonusTokenIndices.push(tokenIndex);
  }
  return state;
}

export function apply(action: GameAction, playerID: string, input: GameState): GameState {
  if (isGameOver(input)) throw new GameError("gameAlreadyOver");
  if (isRoundOver(input)) throw new GameError("roundAlreadyOver");
  if (input.currentPlayerID !== playerID) throw new GameError("notYourTurn");

  const state = structuredClone(input);
  const playerIndex = state.players.findIndex((p) => p.id === playerID);
  if (playerIndex === -1) throw new GameError("notYourTurn");

  switch (action.type) {
    case "takeCamels":
      applyTakeCamels(state, playerIndex);
      break;
    case "takeCard":
      applyTakeCard(state, playerIndex, action.marketCardID);
      break;
    case "exchange":
      applyExchange(state, playerIndex, action.takenMarketCardIDs, action.givenHandCardIDs, action.givenCamelCount);
      break;
    case "sell":
      applySell(state, playerIndex, action.handCardIDs);
      break;
  }

  if (state.players[playerIndex].hand.length > HAND_LIMIT) {
    throw new GameError("handLimitExceeded");
  }

  evaluateRoundEnd(state);

  if (!isRoundOver(state)) {
    const opponent = state.players.find((p) => p.id !== playerID);
    state.currentPlayerID = opponent?.id ?? state.currentPlayerID;
  }

  return state;
}

function refillMarket(state: GameState, count = 5): void {
  const needed = Math.min(count, 5 - state.market.length);
  if (needed <= 0) return;
  const draw = state.drawPile.slice(0, needed);
  state.market.push(...draw);
  state.drawPile = state.drawPile.slice(draw.length);
}

function applyTakeCamels(state: GameState, playerIndex: number): void {
  const camelCards = state.market.filter((c) => c.good === "camel");
  if (camelCards.length === 0) throw new GameError("noCamelsInMarket");
  state.market = state.market.filter((c) => c.good !== "camel");
  state.players[playerIndex].camelCount += camelCards.length;
  refillMarket(state);
}

function applyTakeCard(state: GameState, playerIndex: number, marketCardID: string): void {
  const cardIndex = state.market.findIndex((c) => c.id === marketCardID);
  if (cardIndex === -1) throw new GameError("cardNotInMarket");
  const card = state.market[cardIndex];
  if (card.good === "camel") throw new GameError("cannotTakeCamelViaSingleTake");
  if (state.players[playerIndex].hand.length >= HAND_LIMIT) throw new GameError("handLimitExceeded");

  state.market.splice(cardIndex, 1);
  state.players[playerIndex].hand.push(card);
  refillMarket(state, 1);
}

function applyExchange(
  state: GameState,
  playerIndex: number,
  takenMarketCardIDs: string[],
  givenHandCardIDs: string[],
  givenCamelCount: number
): void {
  const takenCount = takenMarketCardIDs.length;
  const givenCount = givenHandCardIDs.length + givenCamelCount;
  if (takenCount < 2 || takenCount > 5) throw new GameError("exchangeCountOutOfRange");
  if (takenCount !== givenCount) throw new GameError("exchangeCountMismatch");
  if (new Set(takenMarketCardIDs).size !== takenCount) throw new GameError("exchangeCountMismatch");
  if (new Set(givenHandCardIDs).size !== givenHandCardIDs.length) throw new GameError("exchangeCountMismatch");
  if (state.players[playerIndex].camelCount < givenCamelCount) {
    throw new GameError("exchangeInsufficientCamelsInHerd");
  }

  const takenCards: Card[] = [];
  for (const id of takenMarketCardIDs) {
    const card = state.market.find((c) => c.id === id);
    if (!card) throw new GameError("cardNotInMarket");
    if (card.good === "camel") throw new GameError("exchangeCannotIncludeCamelFromMarket");
    takenCards.push(card);
  }

  const givenCards: Card[] = [];
  for (const id of givenHandCardIDs) {
    const card = state.players[playerIndex].hand.find((c) => c.id === id);
    if (!card) throw new GameError("cardNotInHand");
    givenCards.push(card);
  }

  const takenIDs = new Set(takenMarketCardIDs);
  const givenIDs = new Set(givenHandCardIDs);
  state.market = state.market.filter((c) => !takenIDs.has(c.id));
  state.players[playerIndex].hand = state.players[playerIndex].hand.filter((c) => !givenIDs.has(c.id));
  state.players[playerIndex].hand.push(...takenCards);
  state.players[playerIndex].camelCount -= givenCamelCount;

  state.market.push(...givenCards);
  for (let i = 0; i < givenCamelCount; i++) {
    state.market.push({ id: `camel-given-${Date.now().toString(36)}-${i}-${Math.random().toString(36).slice(2)}`, good: "camel" });
  }
}

function applySell(state: GameState, playerIndex: number, handCardIDs: string[]): void {
  if (handCardIDs.length === 0) throw new GameError("emptyAction");
  const cards: Card[] = [];
  for (const id of handCardIDs) {
    const card = state.players[playerIndex].hand.find((c) => c.id === id);
    if (!card) throw new GameError("cardNotInHand");
    cards.push(card);
  }
  const good = cards[0]?.good;
  if (!good || !cards.every((c) => c.good === good)) {
    throw new GameError("sellRequiresCardsOfSameGood");
  }
  const minimum = good === "diamond" || good === "gold" || good === "silver" ? 2 : 1;
  if (cards.length < minimum) throw new GameError("sellBelowMinimumForPreciousGood");

  const ids = new Set(handCardIDs);
  state.players[playerIndex].hand = state.players[playerIndex].hand.filter((c) => !ids.has(c.id));
  state.discard.push(...cards);

  const tokens = takeTokens(state.tokenBank, good, cards.length);
  const won = state.players[playerIndex].wonTokens[good] ?? [];
  state.players[playerIndex].wonTokens[good] = [...won, ...tokens];

  const bonus = drawBonus(state.bonusTokenBank, cards.length);
  if (bonus !== null) {
    state.players[playerIndex].wonBonusTokens.push(bonus);
  }
}

function evaluateRoundEnd(state: GameState): void {
  if (exhaustedStackCount(state.tokenBank) >= 3) {
    state.roundEndReason = "threeStacksExhausted";
  } else if (state.drawPile.length === 0 && state.market.length < 5) {
    state.roundEndReason = "deckExhausted";
  } else {
    return;
  }
  finishRound(state);
}

function finishRound(state: GameState): void {
  const [p0, p1] = state.players;

  let camelBonus0 = 0;
  let camelBonus1 = 0;
  if (p0.camelCount > p1.camelCount) camelBonus0 = 5;
  else if (p1.camelCount > p0.camelCount) camelBonus1 = 5;

  const result0: RoundResult = {
    playerID: p0.id,
    goodsValue: roundGoodsValue(p0),
    bonusValue: roundBonusValue(p0),
    camelBonus: camelBonus0
  };
  const result1: RoundResult = {
    playerID: p1.id,
    goodsValue: roundGoodsValue(p1),
    bonusValue: roundBonusValue(p1),
    camelBonus: camelBonus1
  };
  state.lastRoundResults = [result0, result1];

  const total0 = roundResultTotal(result0);
  const total1 = roundResultTotal(result1);
  if (total0 !== total1) {
    const winnerID = total0 > total1 ? p0.id : p1.id;
    const winner = findPlayer(state, winnerID);
    if (winner) {
      winner.roundsWon += 1;
      if (winner.roundsWon >= 2) {
        state.winnerID = winnerID;
      }
    }
  }
}

File: JaipurKit/Sources/JaipurKit/GameEngine.swift
import Foundation

/// Pure, deterministic Jaipur rules engine. `GameState` is a value type and every
/// mutation goes through `GameEngine.apply`, which validates the action and returns
/// the resulting state (or throws a `GameError`). This makes the engine trivial to
/// unit test and safe to run inside a Firestore transaction (read state, apply,
/// write state) without any hidden mutable global state.
public enum GameEngine {

    // MARK: - Setup

    /// Deals a brand-new game: shuffles the deck, seeds the market with 3 camels
    /// + 2 random cards, and deals 5 cards to each player.
    public static func newGame(
        playerID1: String,
        playerName1: String,
        playerID2: String,
        playerName2: String,
        rng: RandomNumberGenerator = SystemRandomNumberGenerator()
    ) -> GameState {
        var rng = rng
        var players = [
            Player(id: playerID1, displayName: playerName1),
            Player(id: playerID2, displayName: playerName2)
        ]
        let (market, drawPile) = dealMarketAndDrawPile(rng: &rng)

        var remaining = drawPile
        let deal0 = dealStartingHand(&remaining)
        players[0].hand = deal0.hand
        players[0].camelCount = deal0.camelCount
        let deal1 = dealStartingHand(&remaining)
        players[1].hand = deal1.hand
        players[1].camelCount = deal1.camelCount

        let startingPlayer = Bool.random(using: &rng) ? playerID1 : playerID2

        return GameState(
            players: players,
            market: market,
            drawPile: remaining,
            bonusTokenBank: BonusTokenBank.shuffled(using: &rng),
            currentPlayerID: startingPlayer,
            roundNumber: 1
        )
    }

    /// Resets the board for a new round while keeping each player's `roundsWon` tally.
    public static func startNextRound(
        from state: GameState,
        rng: RandomNumberGenerator = SystemRandomNumberGenerator()
    ) -> GameState {
        var rng = rng
        var players = state.players.map { player -> Player in
            var reset = player
            reset.hand = []
            reset.camelCount = 0
            reset.wonTokens = [:]
            reset.wonBonusTokens = []
            reset.revealedBonusTokenIndices = []
            return reset
        }
        let (market, drawPile) = dealMarketAndDrawPile(rng: &rng)
        var remaining = drawPile
        let deal0 = dealStartingHand(&remaining)
        players[0].hand = deal0.hand
        players[0].camelCount = deal0.camelCount
        let deal1 = dealStartingHand(&remaining)
        players[1].hand = deal1.hand
        players[1].camelCount = deal1.camelCount

        let starter = nextRoundStarter(state)

        return GameState(
            players: players,
            market: market,
            drawPile: remaining,
            tokenBank: TokenBank(),
            bonusTokenBank: BonusTokenBank.shuffled(using: &rng),
            currentPlayerID: starter,
            roundNumber: state.roundNumber + 1
        )
    }

    /// House rule: the player who scored lower in the previous round opens the next
    /// one; on an exact tie, starting player alternates by round number.
    private static func nextRoundStarter(_ state: GameState) -> String {
        guard state.lastRoundResults.count == 2 else {
            return state.players[state.roundNumber % 2].id
        }
        let results = state.lastRoundResults
        if results[0].total == results[1].total {
            return state.players[state.roundNumber % 2].id
        }
        return results[0].total < results[1].total ? results[0].playerID : results[1].playerID
    }

    private static func dealMarketAndDrawPile(rng: inout RandomNumberGenerator) -> (market: [Card], drawPile: [Card]) {
        var deck = Deck.freshCards().shuffled(using: &rng)
        let camelIndex = deck.firstIndex { $0.good == .camel }
        precondition(camelIndex != nil, "Deck must contain camels")

        var market: [Card] = []
        var camelsPlaced = 0
        deck.removeAll { card in
            guard card.good == .camel, camelsPlaced < 3 else { return false }
            market.append(card)
            camelsPlaced += 1
            return true
        }
        market.append(contentsOf: deck.prefix(2))
        deck.removeFirst(2)
        return (market, deck)
    }

    /// Deals the next 5 cards off `drawPile` as a starting hand. Any camels
    /// among them go straight to the herd, never into the hand - camels are
    /// never treated as ordinary hand cards at any point in the game, and
    /// the initial deal is the one place that distinction isn't already
    /// enforced by an action's own rules (every in-game action that could
    /// add a camel to a hand is rejected by `apply`).
    private static func dealStartingHand(_ drawPile: inout [Card]) -> (hand: [Card], camelCount: Int) {
        let dealt = Array(drawPile.prefix(5))
        drawPile.removeFirst(dealt.count)
        let hand = dealt.filter { $0.good != .camel }
        return (hand, dealt.count - hand.count)
    }

    /// Flips one of `playerID`'s bonus tokens face-up. Deliberately not gated by
    /// turn order or round/game-over state - either player can flip either
    /// player's tokens at any point once they're won, matching the physical
    /// game's simultaneous reveal at scoring time. Idempotent: flipping an
    /// already-revealed token is a no-op.
    public static func revealBonusToken(_ playerID: String, tokenIndex: Int, in input: GameState) -> GameState {
        var state = input
        guard let playerIndex = state.index(of: playerID) else { return state }
        guard tokenIndex >= 0, tokenIndex < state.players[playerIndex].wonBonusTokens.count else { return state }
        if !state.players[playerIndex].revealedBonusTokenIndices.contains(tokenIndex) {
            state.players[playerIndex].revealedBonusTokenIndices.append(tokenIndex)
        }
        return state
    }

    // MARK: - Actions

    public static func apply(_ action: GameAction, by playerID: String, to input: GameState) throws -> GameState {
        guard !input.isGameOver else { throw GameError.gameAlreadyOver }
        guard !input.isRoundOver else { throw GameError.roundAlreadyOver }
        guard input.currentPlayerID == playerID else { throw GameError.notYourTurn }
        guard let playerIndex = input.index(of: playerID) else { throw GameError.notYourTurn }

        var state = input

        switch action {
        case .takeCamels:
            try applyTakeCamels(&state, playerIndex: playerIndex)
        case .takeCard(let marketCardID):
            try applyTakeCard(&state, playerIndex: playerIndex, marketCardID: marketCardID)
        case .exchange(let taken, let given, let givenCamels):
            try applyExchange(&state, playerIndex: playerIndex, takenMarketCardIDs: taken, givenHandCardIDs: given, givenCamelCount: givenCamels)
        case .sell(let handCardIDs):
            try applySell(&state, playerIndex: playerIndex, handCardIDs: handCardIDs)
        }

        guard state.players[playerIndex].hand.count <= Player.handLimit else {
            throw GameError.handLimitExceeded
        }

        evaluateRoundEnd(&state)

        if !state.isRoundOver {
            state.currentPlayerID = state.players.first { $0.id != playerID }?.id ?? state.currentPlayerID
        }

        return state
    }

    private static func applyTakeCamels(_ state: inout GameState, playerIndex: Int) throws {
        let camelCards = state.market.filter { $0.good == .camel }
        guard !camelCards.isEmpty else { throw GameError.noCamelsInMarket }
        state.market.removeAll { $0.good == .camel }
        state.players[playerIndex].camelCount += camelCards.count
        refillMarket(&state)
    }

    private static func applyTakeCard(_ state: inout GameState, playerIndex: Int, marketCardID: String) throws {
        guard let cardIndex = state.market.firstIndex(where: { $0.id == marketCardID }) else {
            throw GameError.cardNotInMarket
        }
        let card = state.market[cardIndex]
        guard card.good != .camel else { throw GameError.cannotTakeCamelViaSingleTake }
        guard state.players[playerIndex].hand.count < Player.handLimit else {
            throw GameError.handLimitExceeded
        }
        state.market.remove(at: cardIndex)
        state.players[playerIndex].hand.append(card)
        refillMarket(&state, count: 1)
    }

    private static func applyExchange(
        _ state: inout GameState,
        playerIndex: Int,
        takenMarketCardIDs: [String],
        givenHandCardIDs: [String],
        givenCamelCount: Int
    ) throws {
        let takenCount = takenMarketCardIDs.count
        let givenCount = givenHandCardIDs.count + givenCamelCount
        guard (2...5).contains(takenCount) else { throw GameError.exchangeCountOutOfRange }
        guard takenCount == givenCount else { throw GameError.exchangeCountMismatch }
        guard Set(takenMarketCardIDs).count == takenCount else { throw GameError.exchangeCountMismatch }
        guard Set(givenHandCardIDs).count == givenHandCardIDs.count else { throw GameError.exchangeCountMismatch }
        guard state.players[playerIndex].camelCount >= givenCamelCount else {
            throw GameError.exchangeInsufficientCamelsInHerd
        }

        var takenCards: [Card] = []
        for id in takenMarketCardIDs {
            guard let card = state.market.first(where: { $0.id == id }) else { throw GameError.cardNotInMarket }
            guard card.good != .camel else { throw GameError.exchangeCannotIncludeCamelFromMarket }
            takenCards.append(card)
        }

        var givenCards: [Card] = []
        for id in givenHandCardIDs {
            guard let card = state.players[playerIndex].hand.first(where: { $0.id == id }) else {
                throw GameError.cardNotInHand
            }
            givenCards.append(card)
        }

        state.market.removeAll { card in takenMarketCardIDs.contains(card.id) }
        state.players[playerIndex].hand.removeAll { card in givenHandCardIDs.contains(card.id) }
        state.players[playerIndex].hand.append(contentsOf: takenCards)
        state.players[playerIndex].camelCount -= givenCamelCount

        state.market.append(contentsOf: givenCards)
        for _ in 0..<givenCamelCount {
            state.market.append(Card(good: .camel))
        }
    }

    private static func applySell(_ state: inout GameState, playerIndex: Int, handCardIDs: [String]) throws {
        guard !handCardIDs.isEmpty else { throw GameError.emptyAction }
        var cards: [Card] = []
        for id in handCardIDs {
            guard let card = state.players[playerIndex].hand.first(where: { $0.id == id }) else {
                throw GameError.cardNotInHand
            }
            cards.append(card)
        }
        guard let good = cards.first?.good, cards.allSatisfy({ $0.good == good }) else {
            throw GameError.sellRequiresCardsOfSameGood
        }
        guard cards.count >= good.minimumSaleSize else {
            throw GameError.sellBelowMinimumForPreciousGood
        }

        state.players[playerIndex].hand.removeAll { card in handCardIDs.contains(card.id) }
        state.discard.append(contentsOf: cards)

        let tokens = state.tokenBank.takeTokens(for: good, count: cards.count)
        state.players[playerIndex].wonTokens[good, default: []].append(contentsOf: tokens)

        if let bonus = state.bonusTokenBank.drawBonus(forCardsSold: cards.count) {
            state.players[playerIndex].wonBonusTokens.append(bonus)
        }
    }

    private static func refillMarket(_ state: inout GameState, count: Int = 5) {
        let needed = min(count, 5 - state.market.count)
        guard needed > 0 else { return }
        let draw = Array(state.drawPile.prefix(needed))
        state.market.append(contentsOf: draw)
        state.drawPile.removeFirst(draw.count)
    }

    // MARK: - Round / game end

    private static func evaluateRoundEnd(_ state: inout GameState) {
        if state.tokenBank.exhaustedStackCount >= 3 {
            state.roundEndReason = .threeStacksExhausted
        } else if state.drawPile.isEmpty && state.market.count < 5 {
            state.roundEndReason = .deckExhausted
        } else {
            return
        }
        finishRound(&state)
    }

    private static func finishRound(_ state: inout GameState) {
        guard state.players.count == 2 else { return }
        let p0 = state.players[0]
        let p1 = state.players[1]

        var camelBonus0 = 0
        var camelBonus1 = 0
        if p0.camelCount > p1.camelCount {
            camelBonus0 = 5
        } else if p1.camelCount > p0.camelCount {
            camelBonus1 = 5
        }

        let result0 = RoundResult(playerID: p0.id, goodsValue: p0.roundGoodsValue, bonusValue: p0.roundBonusValue, camelBonus: camelBonus0)
        let result1 = RoundResult(playerID: p1.id, goodsValue: p1.roundGoodsValue, bonusValue: p1.roundBonusValue, camelBonus: camelBonus1)
        state.lastRoundResults = [result0, result1]

        if result0.total != result1.total {
            let winnerID = result0.total > result1.total ? p0.id : p1.id
            if let idx = state.index(of: winnerID) {
                state.players[idx].roundsWon += 1
                if state.players[idx].roundsWon >= 2 {
                    state.winnerID = winnerID
                }
            }
        }
    }
}
